계정 로그인, 회원가입 화면 구현

SeungYongSon opened the pull request:
계정 로그인, 회원가입 UI의 기본적인 부분들을 구현했습니다. 
세세한 UI 들은 기능 구현들을 할 때 같이 구현할 것입니다.

gaejuck committed:
[Update] Sign UI

Sign UI 에 기본적인 부분들을 구현했습니다. 세세한 UI 들은 기능 구현을 할 때
같이 구현할 것입니다.

jinusong commented:
일단 오프라인에서 이야기했듯이 Fragment 구조는 바꾸는게 좋을 것 같아요.
나중에 크게 UI가 바뀌는 경우는 없겠지만 만약에 바뀌게 된다면 구조때문에 엉킬 수도 있을 것 같아서 회원가입에 관련된 UI는 전부 화원가입 컨테이너 프래그먼트에 실어주는게 맞다고 생각해요.

그리고 SingleLiveEvent를 Publish & Behavior Subject가 대체하는 것을 고려했어요.
혼자 생각해보기도 하고 종현이형한테과도 상의해 본 결과 SingleLiveEvent가 생명주기를 맞춰서 관리해준다는 것이 장점인데 굳이 생명주기를 벗어나기 위해 Subject를 사용한다는 건 조금 잘 못된 것 같아요. 
나중에 알게 모르게 누수가 발생할 수도 있고 나중에 이벤트를 추가한다고 봤을 때나 생산적으로나 용이한 방법은 아닌 것 같아요. 
그래서 SingleLiveEvent가 낫지 않을까합니다.

jinusong commented on presentation/src/main/java/com/dsm/dms/presentation/ui/fragment/sign/register/RegisterCompleteFragment.kt:
```diff
@@ -0,0 +1,55 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        backPressedGoToMain()
```
onViewCreated에서 BackPress 위임을 처리할 수도 있는데 전 create 전인 onAttach에서 처리하는 편이에요.

SeungYongSon replied:
ㅇㅎ 그렇군요.
하지만 Fragment 구조를 바꾸어서 해당 상황이 없어졌네요ㅎㅎ

jinusong commented on presentation/src/main/res/drawable/ic_key_focused.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="24">
+  <path
+      android:fillColor="@color/colorSignMain"
+      android:pathData="M22,19h-6v-4h-2.68c-1.14,2.42 -3.6,4 -6.32,4 -3.86,0 -7,-3.14 -7,-7s3.14,-7 7,-7c2.72,0 5.17,1.58 6.32,4L24,9v6h-2v4zM18,17h2v-4h2v-2L11.94,11l-0.23,-0.67C11.01,8.34 9.11,7 7,7c-2.76,0 -5,2.24 -5,5s2.24,5 5,5c2.11,0 4.01,-1.34 4.71,-3.33l0.23,-0.67L18,13v4zM7,15c-1.65,0 -3,-1.35 -3,-3s1.35,-3 3,-3 3,1.35 3,3 -1.35,3 -3,3zM7,11c-0.55,0 -1,0.45 -1,1s0.45,1 1,1 1,-0.45 1,-1 - ...
```
vector의 path 데이터들은 path.xml 파일로 옮겨주세요.

SeungYongSon replied:
이런 방법이!

jinusong commented on presentation/src/main/res/layout/fragment_create_account.xml:
```diff
@@ -0,0 +1,109 @@
+            android:orientation="vertical"
+            app:layout_constraintGuide_percent="0.94" />
+
+        <TextView
```
우리는 AndroidX를 쓰고 있으니 TextView, ImageView 등은 androidx.appcompat.widget으로 대체 가능하니까 대체해주세요.

SeungYongSon replied:
넴

jinusong commented on presentation/src/main/res/layout/fragment_register_complete.xml:
```diff
@@ -0,0 +1,90 @@
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical"
```
? orientation 빼주세요

SeungYongSon replied:
아 맞다

jinusong commented on presentation/src/main/res/layout/fragment_register_complete.xml:
```diff
@@ -0,0 +1,90 @@
+                android:layout_height="wrap_content"
+                android:layout_marginTop="36dp"
+                android:gravity="center_horizontal"
+                android:lineSpacingExtra="0dp"
```
의미없는 스페이싱 같아요

SeungYongSon replied:
ㅆㅇㅈ

jinusong commented on presentation/src/main/java/com/dsm/dms/presentation/ui/fragment/sign/register/account/create/CreateAccountFragment.kt:
```diff
@@ -0,0 +1,44 @@
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe {
+                findNavController().navigate(R.id.action_createAccountFragment_to_certifyAccountFragment)
+                createDisposable.dispose()
```
확실히 SingleLiveEvent가 나은 것 같은게 어차피 Event는 main스레드에서 이루어질텐데 계속 스케줄러를 설정하고 dispose 해주는 중복된 코드가 늘어나네요.

SeungYongSon replied:
넴

jinusong commented on presentation/src/main/java/com/dsm/dms/presentation/ui/fragment/sign/register/account/certify/CertifyAccountFragment.kt:
```diff
@@ -0,0 +1,44 @@
+        get() = R.layout.fragment_certify_account
+
+    @Inject
+    lateinit var registerFragment: RegisterFragment
```
흠... 이러면 안될 것 같아요. 나중에 가면 진짜 많이 꼬일 코드에요.
Fragment를 Inject 받아서 그 프레그먼트의 viewModel을 가져다 쓴다는 건 유지보수측면에서는 안좋은 코드인 것 같다는 것이 제 생각입니다.

SeungYongSon replied:
Dagger 에서 RegisterViewModel 자체를 만들어서 Inject 하는 방법으로 바꾸었습니다.
몰론 최고의 해결책은 아닙니다.. 그래서 추후에 더 좋은 방법이 있으면 수정할 것입니다.

SeungYongSon commented:
맞는말이네요. Subject을 사용해서 굳이 생명주기를 안따라서 얻는 이득이 없어보이네요.
그리고 누수를 생각하면 더욱 그렇고요.
아무튼 Subject 부분을 SingleLiveEvent로 바꾸고,
오프라인 때 이야기한 것처럼 프래그먼트 구조를 수정할게요!

신경써주시고 좋은 피드백 해주셔서 감사합니다~

SeungYongSon commented:
> 일단 오프라인에서 이야기했듯이 Fragment 구조는 바꾸는게 좋을 것 같아요.
> 나중에 크게 UI가 바뀌는 경우는 없겠지만 만약에 바뀌게 된다면 구조때문에 엉킬 수도 있을 것 같아서 회원가입에 관련된 UI는 전부 화원가입 컨테이너 프래그먼트에 실어주는게 맞다고 생각해요.
> 
> 그리고 SingleLiveEvent를 Publish & Behavior Subject가 대체하는 것을 고려했어요.
> 혼자 생각해보기도 하고 종현이형한테과도 상의해 본 결과 SingleLiveEvent가 생명주기를 맞춰서 관리해준다는 것이 장점인데 굳이 생명주기를 벗어나기 위해 Subject를 사용한다는 건 조금 잘 못된 것 같아요.
> 나중에 알게 모르게 누수가 발생할 수도 있고 나중에 이벤트를 추가한다고 봤을 때나 생산적으로나 용이한 방법은 아닌 것 같아요.
> 그래서 SingleLiveEvent가 낫지 않을까합니다.

맞는 말이네요.
생각해보니 굳이 Subject 을 사용해서 생명주기를 따르지 않는다고 얻는 이득이 없어보이네요.
그리고 누수를 생각하면 더욱 그렇고요.

아무튼 Subject 부분들을 SingleLiveEvent 로 바꾸고,
오프라인때 이야기 한 것처럼  프래그먼트 구조를 수정할게요.

신경 써주시고 좋은 피드백 해주셔서 고마워요!

gaejuck committed:
[Update] Register Fragment 구조 수정과 피드백 반영

RegisterFragment
ㄴ CreateAccountFragment
ㄴ CetifyAccountFragment
ㄴ RegisterCompleteFragment
위와 같이 Register Fragment 구조를 수정했습니다.
그리고 RegisterFragment 에 있는 Fragment 들은 RegisterFragment 의 RegisterViewModel 을 서로 같이 사용하는 방법으로 사용했습니다.
추후에 각 Fragment 별로 ViewModel 을 사용할 수 있는 방법이 있으면
기존 방법에서 수정 될 수도 있습니다.

그리고 Subject 들을 SingleLiveEvent 로 바꾸었으며,
ic 의 pathData 값들을 path.xml 로 이동시켰습니다.

jinusong commented on presentation/src/main/res/navigation/register_nav_graph.xml:
```diff
@@ -0,0 +1,39 @@
+            app:enterAnim="@anim/nav_default_enter_anim"
+            app:exitAnim="@anim/nav_default_exit_anim"
+            app:popEnterAnim="@anim/nav_default_pop_enter_anim"
+            app:popExitAnim="@anim/nav_default_pop_exit_anim" />
```
navigate로 이동하는거면 혹시 모르니까 popup해주세요

jinusong commented:
컨플릭트 해결해주세요

gaejuck committed:
[Fix] Conflicting 해결

first-feature 와 dev-sign 간의 conflicting 문제를 해결했습니다.
